numba/listobject.py

```diff
@@ -0,0 +1,1089 @@
+            loaded = context.make_optional_value(builder, tl.item_type, item)
+            builder.store(loaded, pout)
+
+        out = builder.load(pout)
+        return context.make_tuple(builder, resty, [status, out])
+
+    return sig, codegen
+
+
+@overload(operator.getitem)
+def impl_getitem(l, index):
+    if not isinstance(l, types.ListType):
+        return
+
+    indexty = INDEXTY
+    itemty = l.item_type
+
+    if index in types.signed_domain:
+        def integer_impl(l, index):
+            index = handle_index(l, index)
+            castedindex = _cast(index, indexty)
+            status, item = _list_getitem(l, castedindex)
+            if status == ListStatus.LIST_OK:
+                return _nonoptional(item)
+            elif status == ListStatus.LIST_ERR_INDEX:
+                raise IndexError("list index out of range")
+            else:
+                raise AssertionError("internal list error during getitem")
+
+        return integer_impl
+
+    elif isinstance(index, types.SliceType):
+        def slice_impl(l, index):
+            newl = new_list(itemty)
+            for i in handle_slice(l, index):
+                newl.append(l[i])
+            return newl
+
+        return slice_impl
+
+    else:
+        raise TypingError("list indices must be signed integers or slices")
+
+
+@intrinsic
+def _list_setitem(typingctx, l, index, item):
+    """Wrap numba_list_setitem
+    """
+    resty = types.int32
+    sig = resty(l, index, item)
+
+    def codegen(context, builder, sig, args):
+        fnty = ir.FunctionType(
+            ll_status,
+            [ll_list_type, ll_ssize_t, ll_bytes],
+        )
+        [l, index, item] = args
+        [tl, tindex, titem] = sig.args
+        fn = builder.module.get_or_insert_function(fnty,
+                                                   name='numba_list_setitem')
+
+        dm_item = context.data_model_manager[titem]
+        data_item = dm_item.as_data(builder, item)
+        ptr_item = cgutils.alloca_once_value(builder, data_item)
+
+        lp = _list_get_data(context, builder, tl, l)
+        status = builder.call(
+            fn,
+            [
+                lp,
+                index,
+                _as_bytes(builder, ptr_item),
+            ],
+        )
+        return status
+
+    return sig, codegen
+
+
+@overload(operator.setitem)
+def impl_setitem(l, index, item):
+    if not isinstance(l, types.ListType):
+        return
+
+    indexty = INDEXTY
+    itemty = l.item_type
+
+    if index in types.signed_domain:
+        def impl_integer(l, index, item):
+            index = handle_index(l, index)
+            castedindex = _cast(index, indexty)
+            casteditem = _cast(item, itemty)
+            status = _list_setitem(l, castedindex, casteditem)
+            if status == ListStatus.LIST_OK:
+                pass
+            elif status == ListStatus.LIST_ERR_INDEX:
+                raise IndexError("list index out of range from c")
```

Comment:
The `IndexError` should have been raised above in `handle_index`, so this should never be reached. I will change this so that anything other than `LIST_OK` at this stage will lead to an `AssertionError`. The same pattern exists above in the `getitem` overload, so will fix there too.

Reply:
Fixed in 275ed9a1a487ef41ab0f8a32b9400c3e380ea332